src/historywork/WriteVerifiedCheckpointHashesWork.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "historywork/WriteVerifiedCheckpointHashesWork.h"
+#include "catchup/VerifyLedgerChainWork.h"
+#include "history/HistoryManager.h"
+#include "historywork/BatchDownloadWork.h"
+#include "ledger/LedgerManager.h"
+#include "ledger/LedgerRange.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+#include "work/ConditionalWork.h"
+#include <Tracy.hpp>
+#include <algorithm>
+#include <fmt/format.h>
+
+namespace stellar
+{
+;
+
+WriteVerifiedCheckpointHashesWork::WriteVerifiedCheckpointHashesWork(
+    Application& app, LedgerNumHashPair rangeEnd, std::string const& outputFile,
+    std::shared_ptr<HistoryArchive> archive)
+    : BatchWork(app, "write-verified-checkpoint-hashes")
+    , mRangeEnd(rangeEnd)
+    , mCurrCheckpoint(rangeEnd.first)
+    , mArchive(archive)
+    , mOutputFile(std::make_shared<std::ofstream>(outputFile))
+{
+    if (mArchive)
+    {
+        CLOG(INFO, "History") << "selected archive " << mArchive->getName();
+    }
+    if (!*mOutputFile)
+    {
+        throw std::runtime_error("error opening output file " + outputFile);
+    }
+    (*mOutputFile) << "[";
+}
+
+WriteVerifiedCheckpointHashesWork::~WriteVerifiedCheckpointHashesWork()
+{
+    // Each line of output made by a VerifyLedgerChainWork has a trailing comma,
+    // and trailing commas are not a valid end of a JSON array; so we terminate
+    // the array here with an entry that does _not_ have a trailing comma (and
+    // identifies an invalid ledger number anyways).
+    (*mOutputFile) << "\n[0, \"\"]\n]\n";
```

Comment:
question about the API: what's the expected result/effects if this command fails? Right now i think we're left with whatever we were able to validate, is that intended? 
* If you want to be super conservative, consider implementing `onSuccess` and `onFailure*` methods. `onSuccess` would finish writing to the file to make it a valid JSON, whereas `onFailure*` would do cleanup: close and remove the file.
* If you want to provide at least some partial result, you could finish writing to the file regardless, though It's probably also worth being a bit more explicit about how the file is reset (on a retry, for example). Right now we're relying on the fact that std::ofstream will truncate any existing file by default, but it's not obvious. So maybe explicitly pass openmode, or add a comment?

Reply:
I want to support partial results so I've done what you suggest in the second part here, implemented a clear open/close pair called at reset as well as construct/destruct, and with an explicit truncate flag.